File: extras/CSharp/Tests/CommandMessenger.IntegrationTests/FirmwareProcessTransport.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.IO;
using System.Threading;
using CommandMessenger.Transport;

namespace CommandMessenger.IntegrationTests
{
    /// <summary>
    /// ITransport that spawns the loopback firmware as a subprocess and communicates
    /// via redirected stdin/stdout pipes. Used for cross-stack integration tests
    /// where we want to exercise both the C# host stack and the C++ firmware stack
    /// over a realistic byte-stream protocol — without any hardware.
    /// </summary>
    public class FirmwareProcessTransport : ITransport
    {
        private readonly string _executablePath;
        private Process _process;
        private Stream _stdin;
        private Stream _stdout;
        private Thread _readerThread;
        private readonly ConcurrentQueue<byte[]> _rxQueue = new();
        private volatile bool _running;

        public event EventHandler DataReceived;

        public FirmwareProcessTransport(string executablePath)
        {
            _executablePath = executablePath ?? throw new ArgumentNullException(nameof(executablePath));
        }

        public bool Connect()
        {
            if (!File.Exists(_executablePath))
                throw new FileNotFoundException("Firmware executable not found", _executablePath);

            var psi = new ProcessStartInfo
            {
                FileName = _executablePath,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            _process = Process.Start(psi);
            if (_process == null) return false;

            _stdin = _process.StandardInput.BaseStream;
            _stdout = _process.StandardOutput.BaseStream;
            _running = true;

            _readerThread = new Thread(ReaderLoop) { IsBackground = true, Name = "FirmwareStdoutReader" };
            _readerThread.Start();

            return true;
        }

        public bool Disconnect()
        {
            _running = false;
            try { _stdin?.Close(); } catch { }
            try
            {
                if (_process != null && !_process.HasExited)
                {
                    if (!_process.WaitForExit(2000))
                        _process.Kill();
                }
            }
            catch { }
            try { _readerThread?.Join(500); } catch { }
            return true;
        }

        public bool IsConnected() => _process != null && !_process.HasExited;

        public byte[] Read()
        {
            if (_rxQueue.TryDequeue(out var data))
                return data;
            return Array.Empty<byte>();
        }

        public void Write(byte[] buffer)
        {
            if (buffer == null || buffer.Length == 0) return;
            _stdin.Write(buffer, 0, buffer.Length);
            _stdin.Flush();
        }

        private void ReaderLoop()
        {
            var buf = new byte[256];
            try
            {
                while (_running)
                {
                    int n = _stdout.Read(buf, 0, buf.Length);
                    if (n <= 0) break; // EOF
                    var chunk = new byte[n];
                    Array.Copy(buf, chunk, n);
                    _rxQueue.Enqueue(chunk);
                    DataReceived?.Invoke(this, EventArgs.Empty);
                }
            }
            catch { /* pipe closed */ }
        }

        public void Dispose()
        {
            Disconnect();
            _process?.Dispose();
        }
    }
}

File: extras/Python/tests/firmware_process_transport.py
"""Subprocess transport for native loopback firmware integration tests."""
from __future__ import annotations

import subprocess
import threading
from collections import deque

from cmd_messenger.transport.transport import Transport


class FirmwareProcessTransport(Transport):
    """Transport that talks to native loopback firmware over stdin/stdout pipes."""

    def __init__(self, executable_path: str) -> None:
        super().__init__()
        self._executable_path = executable_path
        self._process: subprocess.Popen[bytes] | None = None
        self._buffer: deque[bytes] = deque()
        self._lock = threading.Lock()
        self._running = False
        self._reader_thread: threading.Thread | None = None

    def connect(self) -> bool:
        self._process = subprocess.Popen(
            [self._executable_path],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            creationflags=subprocess.CREATE_NO_WINDOW if hasattr(subprocess, "CREATE_NO_WINDOW") else 0,
        )
        self._running = True
        self._reader_thread = threading.Thread(
            target=self._reader_loop,
            name="FirmwareStdoutReader",
            daemon=True,
        )
        self._reader_thread.start()
        return True

    def disconnect(self) -> bool:
        self._running = False
        process = self._process
        if process is None:
            return True
        try:
            if process.stdin is not None:
                process.stdin.close()
        except Exception:
            pass
        try:
            process.wait(timeout=2.0)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait(timeout=2.0)
        if self._reader_thread is not None:
            self._reader_thread.join(timeout=0.5)
        return True

    def is_connected(self) -> bool:
        return self._process is not None and self._process.poll() is None

    def read(self) -> bytes:
        with self._lock:
            if self._buffer:
                return self._buffer.popleft()
        return b""

    def write(self, data: bytes) -> None:
        process = self._process
        if not data or process is None or process.stdin is None:
            return
        process.stdin.write(data)
        process.stdin.flush()

    def dispose(self) -> None:
        self.disconnect()

    def _reader_loop(self) -> None:
        process = self._process
        if process is None or process.stdout is None:
            return
        while self._running and process.poll() is None:
            chunk = process.stdout.read(1)
            if not chunk:
                break
            with self._lock:
                self._buffer.append(chunk)
            self.data_received(self)
